### PR TITLE
Working to eliminate node_modules rewriting

### PR DESCRIPTION
**Pull request by ef4**
This is picking up the work explored in #1374. I'm going to probably land those same code changes, but in a different order so we can more easily keep the tests green through the refactor.

My plan here:

1. Extract reusable API for explicitly asking for rewritten packages, as in https://github.com/embroider-build/embroider/pull/1374/commits/aac1c754f85cfe63f6b4b6240bb647a3b4f6347a.
2. Use that API explicitly everywhere that wants to see rewritten packages.
3. This eliminates the need for MovedPackageCache and lets all package caches get unified as in https://github.com/embroider-build/embroider/pull/1374/commits/d7c17fee90a5c53704cd77b085b2c2fe2bb698ae.
4. Then it should be easier to land the simplified stage1 as in https://github.com/embroider-build/embroider/pull/1374/commits/cc3185e778d2651f183a4db2a163618058bf7521

### CODE REVIEW

**Comment by ef4** [comment]
Note to self: pretty sure we have fully passing tests at 6d54710, it's just that GitHub actions was spazzing out.

File: packages/compat/src/compat-app-builder.ts [hunk 1]
```diff
@@ -1443,9 +1524,9 @@ function stringOrBufferEqual(a: string | Buffer, b: string | Buffer): boolean {
   return false;
 }
 
-const babelFilterTemplate = compile(`
-const { babelFilter } = require(${JSON.stringify(require.resolve('./index.js'))});
-module.exports = babelFilter({{json-stringify skipBabel}}, "{{js-string-escape appRoot}}");
+const babelFilterTemplate = jsHandlebarsCompile(`
+const { babelFilter } = require(${JSON.stringify(require.resolve('@embroider/core'))});
+module.exports = babelFilter({{json-stringify skipBabel}}}, "{{{js-string-escape appRoot}}");
```
Comment (void-mAlex) [inline packages/compat/src/compat-app-builder.ts:1529]: this does not look right, or am I missing something?

Comment (ef4) [reply packages/compat/src/compat-app-builder.ts]: I didn't get the merge conflict right on the first try.

**Comment by ef4** [comment]
Test success count steadily climbing here...

**Comment by ef4** [comment]
This is getting really close. The remaining failures are because the new uses of tree-sync that are introduced turn out to be too aggressive (tree-sync wants to control its entire output directory and will delete anything it doesn't recognize, whereas we're trying to blend several outputs into the same directory). Things like the public assets are getting deleted, and the test are picking up on that.

**Comment by NullVoxPopuli** [comment]
Congrats! Must feel great to finally have this done!

**Comment by jrjohnson** [comment]
This afternoon after the 3.1.0 release we started seeing `Cannot find module './node_modules/@percy/sdk-utils/dist/bundle.js' from '/home/runner/work/common/common/node_modules/.embroider/rewritten-packages/@embroider/synthesized-vendor'` in our [embroider test runs in an addon](https://github.com/ilios/common/actions/runs/5406985232/jobs/9824615075#step:5:38). I haven't dug into it yet, but wanted to mention it quickly in case it helps and it seems like from limited context it is probably coming from this change.

**Comment by Techn1x** [comment]
@jrjohnson I was about to comment the same thing - same `@percy/sdk-utils` package errorr for me too.

I'm using pnpm if that matters - I get the error even after a fresh pnpm - `pnpm store prune && pnpm install --force`

**Comment by ef4** [comment]
@jrjohnson @Techn1x we debugged this today and @mansona is going to prepare the fix. Thanks for reporting.

**Comment by jrjohnson** [comment]
That's fantastic, thanks @ef4 and @mansona!

**Comment by Techn1x** [comment]
Oh boy, when I consume the updates I get past that issue with percy/sdk-utils (thanks so much!!), but end up with a whole different error on build..
```
ERROR in ../rewritten-packages/ember-modal-dialog.fc5b408b/_app_/components/modal-dialog.js 1:0-137
Module not found: Error: Can't resolve '../../../../../.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat' in '/home/boverton/Blake/teacher-ui-client/node_modules/.embroider/rewritten-app/components'
resolve '../../../../../.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat' in '/home/boverton/Blake/teacher-ui-client/node_modules/.embroider/rewritten-app/components'
  using description file: /home/boverton/Blake/teacher-ui-client/node_modules/.embroider/rewritten-app/package.json (relative path: ./components)
    Field 'browser' doesn't contain a valid alias configuration
    No description file found in /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon or above
    no extension
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat doesn't exist
    .wasm
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.wasm doesn't exist
    .mjs
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.mjs doesn't exist
    .js
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.js doesn't exist
    .json
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.json doesn't exist
    .ts
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.ts doesn't exist
    .hbs
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.hbs doesn't exist
    .hbs.js
      Field 'browser' doesn't contain a valid alias configuration
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat.hbs.js doesn't exist
    as directory
      /home/boverton/Blake/.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat doesn't exist
 @ ./assets/teacher-ui-client.js 5276:13-68

webpack 5.88.1 compiled with 1 error in 23875 ms
cleaning up...
Build Error (PackagerRunner) in ../rewritten-packages/ember-modal-dialog.fc5b408b/_app_/components/modal-dialog.js

Module not found: Error: Can't resolve '../../../../../.pnpm/@embroider+macros@1.12.0_@glint+template@1.0.2/node_modules/@embroider/macros/src/addon/es-compat' in '/home/boverton/Blake/teacher-ui-client/node_modules/.embroider/rewritten-app/components'

```

Looks like this issue is related to ember-modal-dialog somehow. When I remove that addon from my app, and all references to it, I get past the error and my app builds and runs.

I'm not sure if maybe I have embroider configured incorrectly or someting - it looks very similar to the error stack encountered here https://github.com/embroider-build/embroider/issues/1371#issuecomment-1463021957 but that could be a red herring.

**Pull request merged**